### DISCUSSION TITLE
S3 storage provider for campaign files

### DISCUSSION THREAD

**Discussion by imilchev** [podkrepi-bg/infrastructure#15] [Ideas]
We already determined some time ago that we will be using S3 bucket(s) for storing campaign-related files. For that we have 2 options - host the S3 storage ourselves or use Amazon's S3.

## Self-hosted S3
We already run Ceph on our cluster and it can create S3 buckets easily with minimal configuration. Of course, storing these files ourselves would mean we have higher storage, CPU and RAM requirements. It also means that we are in control of the data and we do not export it to 3rd parties. I don't know if that can be an issue given that the data itself can be sensitive documents containing personal data.

Hosting it ourselves would also mean that we need to define a backup strategy for the buckets. Because we will need to backup the content of the buckets off-site that would probably mean that we need to use Amazon S3 anyways.

## Amazon S3
We could also go for directly uploading the files to a bucket on Amazon. Cost-wise this will not be any different from a self-hosted solution (assuming that we are going to be backing up our self-hosted buckets to S3). It might mean higher latency but I would not expect this to be a problem. My main concern is if we have to take some extra actions to be GDPR-compliant. If someone from legal can confirm that it would be nice.

**Comment by igoychev** [comment]
Some more reasoning for preferring **Self-hosted S3**:
The organization does not have "secure" budget and does not charge for usage, so using managed services that require regular payments could lock the project out of money. With **Self-hosted S3** we are more agile as it allows running on any hosting providers and even run on private servers donated by the community. 
Also since main team is based in Bulgaria, we thought/dreamed we can have the project running on Bulgarian local servers or hosting providers. This helps GDPR concerns too as the main target users are Bulgarian. The choice of Self-Hosted of course doesn't limit expanding to any geo-location.
